### PR TITLE
Improved toolbar display and selection

### PR DESCRIPTION
**Pull request by jmuharsky**
* Moved toolbars to the right of the selected panel.
* Made toolbar selection dependent on selected panel.
* Moved global buttons (log, options) to the title.
* Restyled the title/header.
* Updated footer styling.
* Updates to color and shading for tab categories.

### CODE REVIEW

File: client/components/container/container-toolbar-directive.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        /** @export */
+        this.insertContainerBeforeSelected = function() {
+          var container = this.dashboardSvc.selectedContainer;
+          var index = this.dashboardSvc.current.model.children.indexOf(container);
+          this.dashboardSvc.addContainerAt(index);
+        };
+
+        /** @export */
+        this.insertContainerAfterSelected = function() {
+          var container = this.dashboardSvc.selectedContainer;
```
Comment (cmccoy) [inline client/components/container/container-toolbar-directive.js]: `insertBefore/After` is accessing some deeply nested attributes of DashboardService - could these functions be moved there?

File: client/components/container/container-toolbar-directive.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+        /** @export */
+        this.moveSelectedContainerUp = function() {
+          var container = this.dashboardSvc.selectedContainer;
+          var containers = this.dashboardSvc.current.model.children;
+          var index = containers.indexOf(container);
+
+          if (index == 0 || containers.length == 0) { return; }
+
+          goog.array.removeAt(containers, index);
```
Comment (cmccoy) [inline client/components/container/container-toolbar-directive.js]: [`goog.array.moveItem`](http://google.github.io/closure-library/api/namespace_goog_array.html#moveItem)?

File: client/components/container/container-toolbar-directive.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+          var container = this.dashboardSvc.selectedContainer;
+          var containers = this.dashboardSvc.current.model.children;
+          var index = containers.indexOf(container);
+
+          if (containers.length == 0 || index == containers.length - 1) {
+            return;
+          }
+
+          goog.array.removeAt(containers, index);
+          goog.array.insertAt(containers, container, index + 1);
```
Comment (cmccoy) [inline client/components/container/container-toolbar-directive.js]: [`goog.array.moveItem`](http://google.github.io/closure-library/api/namespace_goog_array.html#moveItem)?

File: client/components/container/container-toolbar-directive.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+          var container = this.dashboardSvc.selectedContainer;
+          var index = this.dashboardSvc.current.model.children.indexOf(container);
+          this.dashboardSvc.addContainerAt(index);
+        };
+
+        /** @export */
+        this.insertContainerAfterSelected = function() {
+          var container = this.dashboardSvc.selectedContainer;
+          var containers = this.dashboardSvc.current.model.children;
+          var index = containers.indexOf(container);
```
Comment (cmccoy) [inline client/components/container/container-toolbar-directive.js]: Consider asserting that the container was found.

File: client/components/container/container-toolbar-directive.js (hunk 1 above)
Comment (jmuharsky) [inline client/components/container/container-toolbar-directive.js]: Added a TODO #139 to refactor all container code into a single service, and will be done in r24, when the "client" area gets some work.

File: client/components/container/container-toolbar-directive.js (hunk 3 above)
Comment (jmuharsky) [inline client/components/container/container-toolbar-directive.js]: Done.

File: client/components/container/container-toolbar-directive.js (hunk 2 above)
Comment (jmuharsky) [inline client/components/container/container-toolbar-directive.js]: Done.

File: client/components/container/container-toolbar-directive.js (hunk 4 above)
Comment (jmuharsky) [inline client/components/container/container-toolbar-directive.js]: Done; refactored into getSelectedContainerIndex() method.

File: client/components/dashboard/dashboard-toolbar-directive.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+          $window.open(
+              '/dashboard/view?id=' + selectedDashboard.id +
+              '&filename=perfkit_dashboard_' + selectedDashboard.id + '.json')
+        };
+
+        /** @export */
+        this.deleteDashboard = function() {
+          if (!this.dashboard.current.model.id) {
+            console.log('deleteDashboard failed: No id set.');
```
Comment (cmccoy) [inline client/components/dashboard/dashboard-toolbar-directive.js]: `$log.error()`?

Could this be an assert? Is the delete button enabled when `id` is not set?

File: client/components/dashboard/dashboard-toolbar-directive.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            return;
+          }
+
+          if (!window.confirm(
```
Comment (cmccoy) [inline client/components/dashboard/dashboard-toolbar-directive.js]: `$window`?

File: client/components/dashboard/dashboard-toolbar-directive.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+          var promise = this.dashboard.deleteDashboard(
+              this.dashboard.current.model);
+
+          promise.then(angular.bind(this, function(response) {
+            this.openDashboardAdmin();
+          }));
+
+          promise.then(null, angular.bind(this, function(error) {
+            console.log(error.message);
```
Comment (cmccoy) [inline client/components/dashboard/dashboard-toolbar-directive.js]: Please use error service here.

File: client/components/dashboard/dashboard-toolbar-directive.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        ngModel: '='
+      },
+      transclude: false,
+      templateUrl: '/static/components/dashboard/dashboard-toolbar-directive.html',
+      controller: ['$scope', '$window', function($scope, $window) {
+        this.dashboardSvc = dashboardService;
+        this.explorerSvc = explorerService;
+        this.tabSvc = sidebarTabService;
+
+        /** @export */
```
Comment (cmccoy) [inline client/components/dashboard/dashboard-toolbar-directive.js]: Please document methods defined here.

File: client/components/dashboard/dashboard-toolbar-directive_test.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+      var targetElement = actualElement.find(
+          'button.dashboard-copy');
+      expect(targetElement.length).toBe(1);
+
+      spyOn(actualController, 'saveDashboardCopy');
+      targetElement.click();
+      expect(actualController.saveDashboardCopy).toHaveBeenCalled();
+    });
+
+    it('download the current dashboard', function() {
```
Comment (cmccoy) [inline client/components/dashboard/dashboard-toolbar-directive_test.js]: Super nit: download**s**.

File: client/components/explorer/explorer-header-directive.html [hunk 10]
```diff
@@ -1,199 +1,52 @@
...
-  </div>
 
-  <div class="pk-header-section pk-toolbar" ng-hide="ctrl.explorer.model.hideToolbar">
+  <div class="pk-header-toolbar">
+
+    <button type="button" class="btn btn-default"
+            ng-click="ctrl.explorerSvc.showLog()"
+            ng-show="ctrl.errorService.errors.length > 0 &&
+             !(ctrl.explorerSvc.model.code_editor.isOpen &&
+               ctrl.explorerSvc.model.code_editor.selectedMode == 'LOG')">
```
Comment (cmccoy) [inline client/components/explorer/explorer-header-directive.html]: Consider a method for this condition, and introducing a constant for 'LOG'

File: client/components/explorer/sidebar/sidebar-tab-service.js [hunk 11]
```diff
@@ -24,125 +24,99 @@ goog.provide('p3rf.perfkit.explorer.components.explorer.sidebar.SidebarTabServic
...
-  if (this.dashboardService_.selectedWidget) {
-    return this.tabs[this.tabs.length - 1];
-  } else {
-    for (var i=this.tabs.length - 1; i >= 0; --i) {
-      var currentTab = this.tabs[i];
+  SidebarTabService.prototype.getFirstTab = function() {
+    if (this.dashboardService_.selectedWidget) {
+      return this.tabs[0];
+    } else {
+      for (var i=0, len=this.tabs.length; i < len; ++i) {
```
Comment (cmccoy) [inline client/components/explorer/sidebar/sidebar-tab-service.js]: If you're compiling ES6, `for...of` would work well here.

File: client/components/dashboard/dashboard-toolbar-directive.js (hunk 5 above)
Comment (jmuharsky) [inline client/components/dashboard/dashboard-toolbar-directive.js]: Good call.  Changed to assert.

File: client/components/dashboard/dashboard-toolbar-directive.js (hunk 6 above)
Comment (jmuharsky) [inline client/components/dashboard/dashboard-toolbar-directive.js]: Done.

File: client/components/dashboard/dashboard-toolbar-directive.js (hunk 7 above)
Comment (jmuharsky) [inline client/components/dashboard/dashboard-toolbar-directive.js]: Done, and added helper methods to errorService to reduce namespace requirements.

File: client/components/dashboard/dashboard-toolbar-directive.js (hunk 8 above)
Comment (jmuharsky) [inline client/components/dashboard/dashboard-toolbar-directive.js]: Done.

File: client/components/dashboard/dashboard-toolbar-directive_test.js (hunk 9 above)
Comment (jmuharsky) [inline client/components/dashboard/dashboard-toolbar-directive_test.js]: The describe() for this block is "should contain an element to", so "download" is the correct tense here.

**Comment by jmuharsky** [comment]
Addressed comments.  PTAL?

File: client/components/container/container-toolbar-directive.js [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+
+goog.scope(function() {
+  const explorer = p3rf.perfkit.explorer;
+
+
+  /**
+   * See module docstring for more information about purpose and usage.
+   *
+   * @return {Object} Directive definition object.
```
Comment (cmccoy) [inline client/components/container/container-toolbar-directive.js]: `!Object`

File: client/components/container/container-toolbar-directive.js [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+        /**
+         * Returns the index of the selected container.
+         * @return {number} The index of the container, or null if no container
+         *     is selected.
+         */
+        this.getSelectedContainerIndex = function() {
+          if (this.dashboardSvc.selectedContainer === null) {
+            return null;
+          };
```
Comment (cmccoy) [inline client/components/container/container-toolbar-directive.js]: Nit: superfluous `;`

File: client/components/container/container-toolbar-directive.js [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+          var index = this.dashboardSvc.current.model.children.indexOf(
```
Comment (cmccoy) [inline client/components/container/container-toolbar-directive.js]: `let`? You're using `const` above.

File: client/components/container/container-toolbar-directive.js [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+          var index = this.getSelectedContainerIndex();
+          var containers = this.dashboardSvc.current.model.children;
+
+          this.dashboardSvc.removeContainer(
+              this.dashboardSvc.selectedContainer);
+
+          if (containers.length > 0) {
+            var newContainer = null;
+
+            if (index > containers.length - 1) {
```
Comment (cmccoy) [inline client/components/container/container-toolbar-directive.js]: Super-nit: `Math.min(index, containers.length - 1)` would be a little simpler.

File: client/components/dashboard/dashboard-toolbar-directive.js [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+
+goog.scope(function() {
+  const explorer = p3rf.perfkit.explorer;
+
+
+  /**
+   * See module docstring for more information about purpose and usage.
+   *
+   * @return {Object} Directive definition object.
```
Comment (cmccoy) [inline client/components/dashboard/dashboard-toolbar-directive.js]: `!Object`?

File: client/components/container/container-toolbar-directive.js (hunk 12 above)
Comment (jmuharsky) [inline client/components/container/container-toolbar-directive.js]: Done

File: client/components/container/container-toolbar-directive.js (hunk 13 above)
Comment (jmuharsky) [inline client/components/container/container-toolbar-directive.js]: Done.

File: client/components/container/container-toolbar-directive.js (hunk 14 above)
Comment (jmuharsky) [inline client/components/container/container-toolbar-directive.js]: All var's changed to let's.

File: client/components/container/container-toolbar-directive.js (hunk 15 above)
Comment (jmuharsky) [inline client/components/container/container-toolbar-directive.js]: Done.

File: client/components/dashboard/dashboard-toolbar-directive.js (hunk 16 above)
Comment (jmuharsky) [inline client/components/dashboard/dashboard-toolbar-directive.js]: Done.

**Comment by jmuharsky** [comment]
Addressed comments. PTAL?

**Comment by cmccoy** [comment]
LGTM.

**Pull request merged**